docs/stdout-exporter-design.md

```diff
@@ -0,0 +1,223 @@
+# StdoutExporter Design
+
+In strongly typed languages typically there will be 2 separate `Exporter` interfaces, one that accepts spans from a tracer (SpanExporter) and one that accepts metrics (MetricsExporter)
+
+The exporter SHOULD be called with a checkpoint of finished (possibly dimensionally reduced) export records. Most configuration decisions have been made before the exporter is invoked, including which instruments are enabled, which concrete aggregator types to use, and which dimensions to aggregate by.
+
+## Use Cases
+
+Monitoring and alerting systems commonly use the data provided through metric events or tracers, after applying various aggregations and converting into various exposition format. After getting the data, the systems need to be able to see the data. The StdoutExporter will be used here to print data through Stdout, this is seen as a simple exporter where the user doesn’t have the burden of implementing or setting up a protocol dependent exporter.
+
+The StdoutExporter will also be used as a debugging tool for the Metrics API/SDK and Tracing API/SDK which are currently work in progress projects. This exporter will allow contributors to easily diagnose problems when working on the project.
+
+## Push vs Pull Exporter
+
+There are two different versions of exporters: Push and Pull. A Push Exporter pushes the data outwards towards a system, in the case of the StdoutExporter it prints data into Stdout. A Pull Exporter exposes data to some endpoint for another system to grab the data. 
```

Comment:
It'll be vague to tell if an in-memory exporter (an exporter which stores the received data in memory) is push or pull. Not a blocker for this PR though.

Reply:
Is `push` and `pull` the standard terminology used by OT elsewhere?

Can we consider using terms:

- `synchronous` that guarantees the span, context, attributes, being `pushed` (and that could happen without necessarily requiring a copy - it `streams` all info as pass-thru); versus

- `asynchronous` that needs to copy all parameters without, and subsequently allows - either other system to grab the data, OR performs batching and aggregation.

?

Reply:
I got this terminology from my teammates working on the Prometheus exporter. They pointed me to Java which has the terminology here - https://github.com/open-telemetry/opentelemetry-java/tree/master/sdk/src/main/java/io/opentelemetry/sdk/metrics/export